### DISCUSSION TITLE
DELETE method I am receiving 405

### DISCUSSION THREAD

**Discussion by rgraham7503** [hyperjumptech/monika#709] [General]
I'm getting a 405 when trying to issue a DELETE from Monika and I'm not sure why. I can run the exact same curl command from terminal or from postman without any issues.

INFO: 2022-06-16T15:53:50.091Z - Connected to STUN Server. Monika is running from: xx.xx.xx.xx
INFO: 2022-06-16T15:54:01.973Z 1 id:monitor_auth 200 POST https://website.com/v1.0/endpoint/default/token 474ms
INFO: 2022-06-16T15:54:02.776Z 1 id:monitor_auth 201 POST https://website.com/v2.0/Users 794ms
INFO: 2022-06-16T15:54:03.005Z 1 id:monitor_auth 200 GET https://website.com/v2.0/Users?filter=userName+eq+%22{{ response.[1].data.userName}}%22 224ms
WARN: 2022-06-16T15:54:03.493Z 1 id:monitor_auth 405 DELETE https://website.com/v2.0/Users/{{ response.[1].data.id}} 483ms, ALERT: response.status != 200

This is the method under the probe
- method: DELETE
        timeout: 7000
        url: https://website.com/v2.0/Users/{{ response.[1].data.id}}
        headers:
          Authorization: Bearer {{ responses.[0].data.access_token }}
        body: {}
        alerts:
          - query: response.time > 2800
            message: Login API Response time is {{ response.time }} ms, expecting less than 2000>
          - query: response.status != 200
            message: Login API Status code is not 200. Please check the service status!
            
            This is the curl call
            curl --location --request DELETE 'https://website.com/v2.0/Users/xxxxxxxx' \
--header 'Authorization: Bearer xxxxxxxxx'
Returns a response of 204

I dont have access to the logs to see what the DELETE request actually looks like. Is there a way to print out the full request to see what Monika is sending?

**Comment by rgraham7503** [comment]
Ok so I tested hard coding the "id" on the delete call and it works. Now I see whats going on. Is there a way to use a response data in the request URL? 
Like its used in the header? Or is there a way to create a variable that is populated from the response data and use that in the URL? 
/v2.0/Users?filter=userName+eq+%22{{ response.[1].data.userName }}%22  <-- either use this or can I do 
myvar={{ response.[1].data.userName }}
/v2.0/Users?filter=userName+eq+%22$myvar%22

**Comment by rgraham7503** [comment]
Ok so under notifications there is an example of using responses.data in the url. I have tried this and it seems to be literally sending {{ response.[1].data.id }} Also I see differences in examples in the documentation of using "responses" vs "response" I have tried both and neither seem to be working on the url in the probe.

**Reply by sapiderman** [reply]
Hi @rgraham7503, 

For me with a json response.body of

```json
{ 
  status: true,
  dbState: "Connected"
}
```

my alert trigger would be:

```yaml
  - query: response.body.dbState != "Connected"
    subject: NOT HEALTHY
    message: "dying here...."
```        
My desktop alert looks like

![image](https://user-images.githubusercontent.com/964106/174278476-1d34ef12-f897-4579-b9b6-9a5c7b231fbf.png)

**Reply by sapiderman** [reply]
aaaaah i see where the "responses" vs "response" confusion.... That's a typo. Will fix it.

**Reply by rgraham7503** [reply]
@sapiderman 
I am using the chained requests. where I have 4 requests.
1. post to get an Access Token 
2. post to create user <-- uses Authorization: Bearer {{ response.[0].data.access_token }} which works
3. get user to get the id value of newly created user <-- Authorization: Bearer {{ response.[0].data.access_token }} which works 
4. delete newly created user <-- uses Authorization: Bearer {{ response.[0].data.access_token }} AND I can either get the "id" value from response of the second or third request but neither will populate???? {{ response.[2].data.Resources.[0].id }} I have tried {{ response.[2].body.Resources.[0].id }} I have tried a few variations?

Just to see if its a bug or what...I tried putting in {{ responses.[0].data.access_token }} on the DELETE request url to see Monika would populate that value and try to delete a user with the "access token" value, hoping the 405 would go away and just give back a user not found. Nope. This seems like there is a bug where its not populating the values in the url.
So this seems to be broken /v2.0/Users/{{ responses.[1].body.Resources.[].id }}
This is the example I'm using https://monika.hyperjump.tech/guides/examples#multiple-request  Look at "Pass Response Data as Path/Query Parameters"

**Reply by sapiderman** [reply]
Hi @rgraham7503,

Sorry for the late reply. Soo for the {{ response.xxxx.yyy.zzzz }] on the *.yaml,  have you tried putting tick marks '{{ }}'? The {{ response.xxx.yyyy.zzz }} is a javascript feature, and will get a YAML complain when using a YAML parser. 

Also we had an issue using the previous request on the body too... that's fixed here: https://github.com/hyperjumptech/monika/pull/738

Let us know if you still have problems!

Regards,
Budhi

**Comment by rgraham7503** [comment]
Hi @sapiderman I finally got this working! The issue was response vs responses - Notice above what I was using. This is the call that now works 
`/v2.0/Users/{{ responses.[3].body.id }}`
Thank you!! I finally have a full end to end monitor working

**Reply by sapiderman** [reply]
Sweet!

We've improved the documentations too: https://github.com/hyperjumptech/monika/pull/780
Hope it would remove any future confusions. Hats off to all your feedbacks.

We've also got a ticket on improving how we display responses and what we send out (such as from {{ response.xx.xx }}. It would be nice if it resolved properly, say with a --verbose flag.

happy to get all your feedbacks and questions!
